### ISSUE TITLE
skim: use-of-uninitialized-value when using the interactive history

### ISSUE THREAD

**Issue by Algunenano** [ClickHouse/ClickHouse#59599]
* Open `clickhouse local` with a MSAN build
* Write any query
* Ctrl+R to open the interactive history, select the previous query

```
==650265==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x555598cd494a in cxxbridge1$string$ptr (/mnt/ch/ClickHouse/build_msan/programs/clickhouse+0x4378094a) (BuildId: debf3c35c79878593ff4d8a4d91987ee75e4717a)
    #1 0x55559527c206 in auto std::__1::ranges::__data::__fn::operator()[abi:v15000]<rust::cxxbridge1::String const&>(rust::cxxbridge1::String const&) const build_msan/./contrib/llvm-project/libcxx/include/__ranges/data.h:57:18
    #2 0x55559527c206 in std::__1::basic_string_view<char, std::__1::char_traits<char>>::basic_string_view[abi:v15000]<rust::cxxbridge1::String const&>(rust::cxxbridge1::String const&) build_msan/./contrib/llvm-project/libcxx/include/string_view:335:46
    #3 0x55559527c206 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::basic_string<rust::cxxbridge1::String, void>(rust::cxxbridge1::String const&) build_msan/./contrib/llvm-project/libcxx/include/string:2192:24
    #4 0x55559527c206 in DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10::operator()(char32_t) const build_msan/./src/Client/ReplxxLineReader.cpp:426:25
    #5 0x55559527c206 in decltype(std::declval<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&>()(std::declval<char32_t>())) std::__1::__invoke[abi:v15000]<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t>(DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #6 0x55559527c206 in replxx::Replxx::ACTION_RESULT std::__1::__invoke_void_return_wrapper<replxx::Replxx::ACTION_RESULT, false>::__call<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t>(DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/invoke.h:470:16_rust_skim_rust v0.1.0 (/mnt/ch/ClickHouse/build_msan/rust/skim)
    #7 0x55559527c206 in std::__1::__function::__default_alloc_func<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10, replxx::Replxx::ACTION_RESULT (char32_t)>::operator()[abi:v15000](char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #8 0x55559527c206 in replxx::Replxx::ACTION_RESULT std::__1::__function::__policy_invoker<replxx::Replxx::ACTION_RESULT (char32_t)>::__call_impl<std::__1::__function::__default_alloc_func<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10, replxx::Replxx::ACTION_RESULT (char32_t)>>(std::__1::__function::__policy_storage const*, char32_t) build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16
    #9 0x5555a025e9e8 in std::__1::__function::__policy_func<replxx::Replxx::ACTION_RESULT (char32_t)>::operator()[abi:v15000](char32_t&&) const build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #10 0x5555a025e9e8 in std::__1::function<replxx::Replxx::ACTION_RESULT (char32_t)>::operator()(char32_t) const build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #11 0x5555a025e9e8 in replxx::Replxx::ReplxxImpl::get_input_line() build_msan/./contrib/replxx/src/replxx_impl.cxx:1417:11
    #12 0x5555a025da35 in replxx::Replxx::ReplxxImpl::input(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) build_msan/./contrib/replxx/src/replxx_impl.cxx:662:8
    #13 0x5555a0213abe in replxx::Replxx::input(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) build_msan/./contrib/replxx/src/replxx.cxx:150:18
    #14 0x5555952750fe in DB::ReplxxLineReader::readOneLine(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) build_msan/./src/Client/ReplxxLineReader.cpp:482:29
    #15 0x555595201f97 in DB::LineReader::readLine(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) build_msan/./src/Client/LineReader.cpp:148:26
    #16 0x5555950de5a9 in DB::ClientBase::runInteractive() build_msan/./src/Client/ClientBase.cpp:2477:24
    #17 0x555577eb584a in DB::LocalServer::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_msan/./programs/local/LocalServer.cpp
    #18 0x5555a003d38b in Poco::Util::Application::run() build_msan/./base/poco/Util/src/Application.cpp:315:8
    #19 0x555577ed6125 in mainEntryClickHouseLocal(int, char**) build_msan/./programs/local/LocalServer.cpp:954:20
    #20 0x55555e06403e in main build_msan/./programs/main.cpp:515:12                                                                                                                                                                                         
    #21 0x7ffff7dcdccf in __libc_start_call_main /usr/src/debug/glibc/glibc/csu/../sysdeps/nptl/libc_start_call_main.h:58:16                                                                                                                                 
    #22 0x7ffff7dcdd89 in __libc_start_main /usr/src/debug/glibc/glibc/csu/../csu/libc-start.c:360:3lock;                                                                                                                                                    
    #23 0x55555dfd492d in _start (/mnt/ch/ClickHouse/build_msan/programs/clickhouse+0x8a8092d) (BuildId: debf3c35c79878593ff4d8a4d91987ee75e4717a)                                                                                                           
                                                                                               
  Uninitialized value was created by an allocation of 'ref.tmp19.i.i.i.i' in the stack frame                                                                                                                                                                 
    #0 0x55559527c1d8 in DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10::operator()(char32_t) const build_msan/./src/Client/ReplxxLineReader.cpp:426:37                                                                                                                                        
    #1 0x55559527c1d8 in decltype(std::declval<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&>()(std::declval<char32_t>())) std::__1::__invoke[abi:v15000]<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t>(DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23  
    #2 0x55559527c1d8 in replxx::Replxx::ACTION_RESULT std::__1::__invoke_void_return_wrapper<replxx::Replxx::ACTION_RESULT, false>::__call<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t>(DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10&, char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/invoke.h:470:16
    #3 0x55559527c1d8 in std::__1::__function::__default_alloc_func<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10, replxx::Replxx::ACTION_RESULT (char32_t)>::operator()[abi:v15000](char32_t&&) build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #4 0x55559527c1d8 in replxx::Replxx::ACTION_RESULT std::__1::__function::__policy_invoker<replxx::Replxx::ACTION_RESULT (char32_t)>::__call_impl<std::__1::__function::__default_alloc_func<DB::ReplxxLineReader::ReplxxLineReader(DB::LineReader::Suggest&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::vector<char const*, std::__1::allocator<char const*>>, std::__1::vector<char const*, std::__1::allocator<char const*>>, char const*, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<replxx::Replxx::Color, std::__1::allocator<replxx::Replxx::Color>>&)>)::$_10, replxx::Replxx::ACTION_RESULT (char32_t)>>(std::__1::__function::__policy_storage const*, char32_t) build_msan/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16

SUMMARY: MemorySanitizer: use-of-uninitialized-value (/mnt/ch/ClickHouse/build_msan/programs/clickhouse+0x4378094a) (BuildId: debf3c35c79878593ff4d8a4d91987ee75e4717a) in cxxbridge1$string$ptr
Exiting
```


Doesn't look related to https://github.com/ClickHouse/ClickHouse/pull/59447
cc @azat

**Comment by alexey-milovidov** [comment]
It is alright if not all the Rust code or its wrapper is instrumented for MSan.
Add `__msan_unpoison` somewhere.

**Comment by azat** [comment]
It actually should be instrumented since https://github.com/ClickHouse/ClickHouse/pull/50541

As for the original problem I cannot reproduce it, one guess is that maybe not all code had been instrumented...

**Comment by Algunenano** [comment]
Here is what I've tested:
* Downloaded a build from master (https://s3.amazonaws.com/clickhouse-test-reports/0/0c35492ed33333a2e39f57d8e3c196d1fd22ae22/clickhouse_build_check/report.html) and it does not reproduce it, so it might be something different from my build.
* A clean local build. Still reproduced it.
* Reverted 65cfbaaa. It fixed it locally.
* Uninstalled all rust targets (both stable and nightly) except `nightly-2023-07-04-x86_64-unknown-linux-gnu` (it was the default). Fixed it too.

So I have no idea what was going on, but leaving only the `nightly-2023-07-04-x86_64-unknown-linux-gnu` toolchain removed the problem, so I'm going to assume something was wrong with my installation (wrong toolchain being used to build some files?).
